lib/std/unicode.zig

```diff
@@ -1,17 +1,67 @@
+const std = @import("std");
 const assert = std.debug.assert;
 const testing = std.testing;
 const mem = std.mem;
 
+// While these are the errors, the return types of the functions cannot be
+// set because the ErrorSet type is too buggy in stage1.
+pub const Utf8Error = UnicodeError || error{
+    Utf8ShortChar,
+    Utf8OverlongEncoding,
+    Utf8InvalidStartByte,
+};
+
+pub const UnicodeError = error{
+    UnicodeSurrogateHalf,
+    UnicodeCodepointTooLarge,
+};
+
+// http://www.unicode.org/versions/Unicode6.0.0/ch03.pdf - page 94
+//
+// Table 3-7. Well-Formed UTF-8 Byte Sequences
+//
+// +--------------------+------------+-------------+------------+-------------+
+// | Code Points        | First Byte | Second Byte | Third Byte | Fourth Byte |
+// +--------------------+------------+-------------+------------+-------------+
+// | U+0000..U+007F     | 00..7F     |             |            |             |
+// +--------------------+------------+-------------+------------+-------------+
+// | U+0080..U+07FF     | C2..DF     | 80..BF      |            |             |
+// +--------------------+------------+-------------+------------+-------------+
+// | U+0800..U+0FFF     | E0         | A0..BF      | 80..BF     |             |
+// +--------------------+------------+-------------+------------+-------------+
+// | U+1000..U+CFFF     | E1..EC     | 80..BF      | 80..BF     |             |
+// +--------------------+------------+-------------+------------+-------------+
+// | U+D000..U+D7FF     | ED         | 80..9F      | 80..BF     |             |
+// +--------------------+------------+-------------+------------+-------------+
+// | U+E000..U+FFFF     | EE..EF     | 80..BF      | 80..BF     |             |
+// +--------------------+------------+-------------+------------+-------------+
+// | U+10000..U+3FFFF   | F0         | 90..BF      | 80..BF     | 80..BF      |
+// +--------------------+------------+-------------+------------+-------------+
+// | U+40000..U+FFFFF   | F1..F3     | 80..BF      | 80..BF     | 80..BF      |
+// +--------------------+------------+-------------+------------+-------------+
+// | U+100000..U+10FFFF | F4         | 80..8F      | 80..BF     | 80..BF      |
+// +--------------------+------------+-------------+------------+-------------+
+
+// This accepts u32 instead of u21 because it is much more common to work with
+// u32s (that is what the CPU is already working with).
+pub fn isValidUnicode(c: u32) !void {
+    switch (c) {
+    0x0000...0xd7ff => {},
+    0xd800...0xdfff => return error.UnicodeSurrogateHalf,
+    0xe000...0x10ffff => {},
+    0x110000...0xffffffff => return error.UnicodeCodepointTooLarge,
+    }
+}
+
 /// Returns how many bytes the UTF-8 representation would require
-/// for the given codepoint.
-pub fn utf8CodepointSequenceLength(c: u21) !u3 {
+/// for the given codepoint. Taking a u21 here would add no safety, as we
+/// already have to check the upper bound, which is smaller than maxInt(u21).
```

Comment:
Zig in general prefers the tighter integer type; in future its likely to be replaced with a ranged integer (see #3806)

Reply:
Yeah but this is a function entry, and if LLVM actually compiles this as a function it will mask the high bits. I can leave it and look into a LLVM patch however.